### PR TITLE
No shrinking for BackdropFilter's cull rect

### LINKED ISSUE #29070: No shrinking for BackdropFilter's cull rect
Currently, it's easy to misunderstand how `BackdropFilter` works because its area of effect depends on Skia's cull rect. Skia views that cull rect only as a performance optimization hint so it can freely modify it to any rect that encloses the child. Therefore it's hard to predict the cull rect's exact size.

One example of such confusion is
```
BackdropFilter(
  child: Container(width: 200, height: 200, child: Text('Hello')),
  ...
);
```

Many expect the filter to be applied to the full container of size 200x200. Instead, Skia only sees a `drawText` call which has an small area of 'Hello', so it reduces the filter area to just around that 'Hello' text.

By fixing this this issue, Flutter will always set `BackdropFilter`'s area of effect to fill its parent/ancestor clip (full screen if unclipped). This is more in line with our clip behaviors (no clip by default), but it doesn't directly achieve a 200x200 filtered area using the code above.

In order to get the filtered area of 200x200 when this issue is fixed, one has to write the following code

```
ClipRect(
  child: BackdropFilter(
    child: Container(width: 200, height: 200, child: Text('Hello')),
  )
);
```

If this breaks your app, wrap the `BackdropFilter` with a `ClipRect`  (or some other clip of your choice).

Related issues: https://github.com/flutter/flutter/issues/10148 (BackdropFilter only shrinks when there's a child text widget), https://github.com/flutter/flutter/issues/16148 (BackdropFilter doesn't match the size of of its child).

To fully fix this issue, we have a 3-step plan to accommodate our golden tests infrastructure.

1. Land https://github.com/flutter/flutter/pull/28174 with the transparent paint hack to change the behavior with pure framework (flutter/flutter) changes.
2. Modify `BackdropFilterLayer` in flutter/engine to send a nullptr bounds (instead of `paint_bounds`) to `SaveLayerRec` in https://github.com/flutter/engine/blob/master/flow/layers/backdrop_filter_layer.cc#L19
3. Wait for the engine change to roll into the framework; then remove the transparent paint hack.

Such plan ensures that our engine roll won't be blocked by a golden image mismatch.

### PR DESCRIPTION
**Pull request by liyuqian**
Fixes https://github.com/flutter/flutter/issues/29070

This will be a breaking change. Our old behavior may generate confusions
for a sample app like our added golden test: Skia will shrink the cull
rect (and thus the filtered area) to the text. The new behavior will
fill the BackdropFilter to its parent/ancestor clip. This is more
in line with our clip behaviors (no clip by default).

If this breaks your app, wrap the BackdropFilter with a ClipRect (or some other clip of your choice).

[wip] The golden images are not uploaded yet. I'll wait for the initial
round of review to approve the golden test before uploading them.

The correct golden image:
backdrop_filter_test cull_rect 1

The incorrect golden image before this PR:
backdrop_filter_test cull_rect 1 wrong

### CODE REVIEW

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (chinmaygarde) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Nit: s/painting/paintingCallback

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (chinmaygarde) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: s/shrinked/shrunk

**Review by chinmaygarde — APPROVED** [review]

File: packages/flutter/test/widgets/backdrop_filter_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/backdrop_filter_test.dart]: Remove this line?

Comment (liyuqian) [reply packages/flutter/test/widgets/backdrop_filter_test.dart]: Done. Thanks!

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (liyuqian) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Done.

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (liyuqian) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Done. Thanks!

**Comment by goderbauer** [comment]
Can you document on BackdropFilter that it may potentially need to be wrapped in a ClipRect?

**Comment by liyuqian** [comment]
CC @Hixie for the breaking change. Since there's no user complaining yet, we may wait a while and send an ask/announcement to flutter-dev before landing this.

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (xster) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Can't this be statically defined? Instead of instantiating a new closure on each paint call.

**Comment by xster** [comment]
LGTM \o/

**Comment by Hixie** [comment]
> Since there's no user complaining yet, we may wait a while and send an ask/announcement to flutter-dev before landing this.

We should definitely do the breaking change policy if we land this, regardless of who is complaining.

**Comment by Hixie** [comment]
This change should have a corresponding documentation change. Either fixing what we say now, or adding documentation to explain what's going on. We need examples of how to get the effect we want, ideally with actual diagrams (see the assets-for-api-docs repo for how to make diagrams) and sample code.

**Comment by spkersten** [comment]
This took us a while to figure out initially and now we are adding a container with `Color.transparant` background around most of our `BackdropFilter`s. Not a deal breaker of course, but slightly inconvenient.

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (liyuqian) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: If I make a static, I can no longer call `super.paint` which is non-static?

Don't worry, I'll remove this hacky callback soon. See my 3-step plan in https://github.com/flutter/flutter/issues/29070#issue-418990863

**Comment by liyuqian** [comment]
Doc updated. Added issue https://github.com/flutter/flutter/issues/29070 with the breaking change announcement. Please review the updated PR.

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (xster) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: I meant make this function statically defined, not make this the function static. 

e.g.

```dart
void paintingCallback(PaintingContext context, Offset offset) {
  blah
}

void paint(PaintingContext context, Offset offset) {
  context.pushLayer(BackdropFilter(filter: _filter), paintingCallback, offset);
}
```

Comment (liyuqian) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Oh, great, that's a good idea. Thanks!

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: nit: remove extra blank line?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/proxy_box.dart
Comment (liyuqian) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Done. Thanks!

**Comment by liyuqian** [comment]
@dnfield after your PhysicalShapeLayer revert, my golden version is having a conflict since my golden patch is after your golden patch. Should I wait for your reland before merging this? (It seems that no new goldens can be updated unless you either reland the PR, or revert the golden commit that includes the PhysicalShapeLayer fix.)

**Comment by dnfield** [comment]
If you can wait for the engine roll to land that would be idea.

**Comment by dnfield** [comment]
Either that, or take the same goldens.version as the engine roll has here: https://github.com/flutter/flutter/pull/29363

**Comment by liyuqian** [comment]
I'll wait for the manual roll :)

**Pull request merged**
